### DISCUSSION TITLE
data augmentation with tf datasets

### DISCUSSION THREAD

**Discussion by metzm** [ctu-geoforall-lab-projects/phd-pesek-2024#50] [Ideas]
Please have a look at [my feature branch](https://github.com/metzm/phd-pesek-2022/tree/augment_tfdataset) for some wild hacks to your code to 

1. create on-disk stored TensorFlow datasets for the training and validation data
2. use the Tensorflow training dataset to add any standard data augmentation methods like some `tf.keras.layers.Random*()` layers

Initial results are not yet satisfying regarding the loss graphs, but maybe this feature branch can provide some inspiration for better data augmentation, improvement of the loss graphs and avoiding overfitting.

The most important changes are the new fn [get_tf_dataset()](https://github.com/metzm/phd-pesek-2022/blob/augment_tfdataset/src/cnn_lib.py#L1156) which creates a tf dataset from the training and validation data, and the new [Augment](https://github.com/metzm/phd-pesek-2022/blob/augment_tfdataset/src/cnn_lib.py#L22) class where various data augmentation methods can be easily added.

**Comment by pesekon2** [comment]
@metzm: That's actually great! The dataset creation was created in quite a tentative, temporary way, and it just stayed that way. I experimented with `tf.data.Dataset.from_tensor_slices`, but it was broken when I started writing it (`tensorflow==2.1.something`). Thanks for looking at it.

Once you find the results satisfying, a PR from your feature branch will be more than appreciated.

**Comment by metzm** [comment]
No satisfying results yet, despite lots of experimenting, the validation loss is still much higher than the training loss. However, I got some small improvements with
* learning rate reduction
* a smaller tensor shape, e.g. (128, 128)
* reduction of the number of filters from 64 to 32 to get something like a tinyunet, apparently less prone to overfitting
* data augmentation, but not all possible data augmentation methods improve the loss graphs, see code comments

For faster experimenting with different parameters (hyperparameter tuning), I created and saved tf datasets beforehand and load them if existing instead of creating them each time anew. This speeds up training and particularly fine-tuning and seems to reduce memory consumption.

**Comment by pesekon2** [comment]
> learning rate reduction

The default learning rate for `Adam` optimizer is 0.001 which seemed already quite low to me, but if you've got improvements with even a lower one, then it is worth it. Maybe we should make `learning_rate` a parameter of the `create_model()` function?

> a smaller tensor shape, e.g. (128, 128)

This is weird. Do you have any idea why? Have you checked the detections manually afterwards (it is possible that the lower loss was only because of fewer problematic pixels in the picture)? 

> reduction of the number of filters from 64 to 32 to get something like a tinyunet, apparently less prone to overfitting

Parameter `nr_filters` of the function `create_model()` should be doing this, right?  
Tinyunet should be great especially for smaller datasets. U-Net can handle more complexities, but if the data amount is not sufficient or the task is not so complex, it can be surely more prone to overfitting than tinyunet.

> For faster experimenting with different parameters (hyperparameter tuning), I created and saved tf datasets beforehand and load them if existing instead of creating them each time anew. This speeds up training and particularly fine-tuning and seems to reduce memory consumption.

+1. That was done before too, but the `tf.Dataset` approach didn't work for me. I support this approach very much.
